addons/website/models/website.py

```diff
@@ -608,30 +610,58 @@ def _render_snippet(key):
             xml_translate(terms.append, render)
             placeholders = [_compute_placeholder(term) for term in terms]
 
+            if text_must_be_translated_for_openai:
+                nonlocal nb_terms_translated
+                nonlocal nb_terms_total
```

Comment:
`nonlocal` is rare in the code base (about ten occurrences). We need it because we write/read a variable defined in the parent function. The alternative would be to have arrays of terms (total and translated) and take into account the size of these arrays to calculate the translation ratio.

Reply:
You don't need that because the current implementation seems wrong :grimacing: You don't cache those values here :grimacing: Here it means that if you render the same snippet twice, you don't increase the total terms. It could be arguably ok but it means that if you use 9 times a very poorly translated snippet, combined with 9 nicely translated other ones, you will have a coverage of ~90% while it is in fact ~50%. What you need here is including the nb translated and total in the `data` variable here (the cache) and not increase the external total in here (but outside, where you received the values as the return value of `_render_snippet`).

Reply:
If the same text appears several times, it will only appear once in `generated_content`. See => `generated_content[placeholder] = ''`.  For OpenAI to give a good result, +80% of the terms sent to it must be translated. This is different from saying that 80% of the theme's text must be translated. This nuance is important, as there is often repetition of text in themes.

> if you use 9 times a very poorly translated snippet, combined with 9 nicely translated other ones, you will have a coverage of ~90% while it is in fact ~50%. 

So in this case OpenAI will receive 90% of placeholder in the correct lang.

Maybe I can add a comment to explain that difference between theme translation ratio and placeholders translation ratio.

Reply:
Indeed, I was wrong :) But then all the implementation in this PR were wrong no? If you count the translated terms at each snippet rendering extraction, you may have the same text counted twice in different snippets while it will only be sent only once to IAP? Basically if a snippet contains another with a t-snippet-call, it's wrong? What about doing this then: instead of counting the terms at each snippet rendering, build an object like `generated_content` => `translated_content[from] = to`. That way if a term is repeated multiple times, you will only count it once. And counting will be very easy at the end: it will be number of from=to in `translated_content`  over the  size of `translated_content`.